update decimal schema converter

veliuenal opened the pull request:
## Overview
This change introduces sink-side controls for how DECIMAL types are represented in BigQuery at both schema-generation and value-conversion time:

New handling modes for fixed-scale Kafka Connect Decimal and Debezium VariableScaleDecimal:
- RECORD (previously discussed as “NONE” in review): write as a STRUCT { scale, value } (Debezium-style payload)
- FLOAT (default; backward compatible): write as BigQuery FLOAT64
- NUMERIC: write as BigQuery NUMERIC with scale/precision awareness
- BIGNUMERIC: write as BigQuery BIGNUMERIC with scale/precision awareness
- Default remains FLOAT to avoid a breaking change

## What Changed

* Configurable decimal handling:
We’ve added a DecimalHandlingMode enum (RECORD, FLOAT, NUMERIC, BIGNUMERIC) to BigQuerySinkConfig so you can choose exactly how decimals land in BigQuery. The setting is case-insensitive, so you don’t have to worry about capitalization when configuring it.

* Record conversion logic:
BigQueryRecordConverter now keeps separate modes for Kafka Decimal and Debezium VariableScaleDecimal and applies them in convertLogical. For Kafka Decimal, RECORD produces a {scale,value} struct, NUMERIC/BIGNUMERIC preserves the original BigDecimal, and FLOAT (the default) converts to a double. Debezium’s VariableScaleDecimal follows the same mapping, yielding a struct, double, or BigDecimal as configured.

* Schema conversion logic:
BigQuerySchemaConverter checks that Kafka Decimal uses a BYTES base type and then builds the BigQuery field according to the selected mode. RECORD emits a nested {scale,value} RECORD; NUMERIC/BIGNUMERIC delegates to the logical converter and logs a warning if precision/scale exceed BigQuery limits; and FLOAT (default) uses a FLOAT64 field. VariableScaleDecimal is handled with the same approach to keep schema and value conversion in sync.

* Logical converter internals:
KafkaLogicalConverters.DecimalConverter still targets LegacySQLTypeName.NUMERIC for its builder and now exposes getFieldBuilder, which lets us propagate precision and scale from schema parameters into BigQuery field definitions when NUMERIC/BIGNUMERIC is chosen. It also enforces BYTES encoding for Kafka Decimal, ensuring we decode consistently before producing either a BigDecimal or the struct form.

* Tests:
New unit tests cover all modes end-to-end. 
testKafkaDecimalNumericMode verifies that NUMERIC produces a BigDecimal in the output map.
testKafkaDecimalRecordMode checks the {scale,value} struct for RECORD. 
testKafkaDecimalDefaultFloat confirms FLOAT (and the default) yields a Double. 
Schema tests mirror these scenarios, including default FLOAT mapping and explicit NUMERIC/BIGNUMERIC with precision/scale propagation.


## Tests

```
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.311 s - in com.wepay.kafka.connect.bigquery.write.storage.StorageWriteApiWriterTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.04 s - in com.wepay.kafka.connect.bigquery.write.storage.BigQueryBuilderTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.009 s - in com.wepay.kafka.connect.bigquery.write.storage.BigQueryWriteSettingsBuilderTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 s - in com.wepay.kafka.connect.bigquery.write.storage.GcsBuilderTest
Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.106 s - in com.wepay.kafka.connect.bigquery.write.storage.StorageWriteApiBatchApplicationStreamTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 s - in com.wepay.kafka.connect.bigquery.write.storage.StorageApiBatchModeHandlerTest
Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.037 s - in com.wepay.kafka.connect.bigquery.write.storage.StorageWriteApiDefaultStreamTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.133 s - in com.wepay.kafka.connect.bigquery.write.row.BigQueryWriterTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.012 s - in com.wepay.kafka.connect.bigquery.write.row.GcsToBqWriterTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.229 s - in com.wepay.kafka.connect.bigquery.GcpClientBuilderProjectTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.027 s - in com.wepay.kafka.connect.bigquery.config.GcsBucketValidatorTest
Tests run: 23, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.015 s - in com.wepay.kafka.connect.bigquery.config.BigQuerySinkConfigTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in com.wepay.kafka.connect.bigquery.config.PartitioningModeValidatorTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in com.wepay.kafka.connect.bigquery.config.MultiPropertyValidatorTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 s - in com.wepay.kafka.connect.bigquery.config.CredentialsValidatorTest
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 s - in com.wepay.kafka.connect.bigquery.config.StorageWriteApiValidatorTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in com.wepay.kafka.connect.bigquery.config.PartitioningTypeValidatorTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in com.wepay.kafka.connect.bigquery.utils.PartitionedTableIdTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in com.wepay.kafka.connect.bigquery.utils.FieldNameSanitizerTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in com.wepay.kafka.connect.bigquery.ErrantRecordHandlerTest
Tests run: 23, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.167 s - in com.wepay.kafka.connect.bigquery.BigQuerySinkTaskTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.125 s - in com.wepay.kafka.connect.bigquery.BigQueryStorageApiBatchSinkTaskTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in com.wepay.kafka.connect.bigquery.BigQuerySinkConnectorTest
Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.033 s - in com.wepay.kafka.connect.bigquery.MergeQueriesTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.112 s - in com.wepay.kafka.connect.bigquery.BigQueryStorageApiSinkTaskTest
Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 s - in com.wepay.kafka.connect.bigquery.exception.BigQueryStorageWriteApiErrorResponsesTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in com.wepay.kafka.connect.bigquery.exception.BigQueryErrorResponsesTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in com.wepay.kafka.connect.bigquery.exception.BigQueryStorageWriteApiConnectExceptionTest
Tests run: 39, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.048 s - in com.wepay.kafka.connect.bigquery.SchemaManagerTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.027 s - in com.wepay.kafka.connect.bigquery.GcsToBqLoadRunnableTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in com.wepay.kafka.connect.bigquery.convert.KafkaDataConverterTest
Tests run: 27, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.063 s - in com.wepay.kafka.connect.bigquery.convert.BigQuerySchemaConverterTest
Tests run: 29, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 s - in com.wepay.kafka.connect.bigquery.convert.BigQueryRecordConverterTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in com.wepay.kafka.connect.bigquery.convert.logicaltype.KafkaLogicalConvertersTest
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 s - in com.wepay.kafka.connect.bigquery.convert.logicaltype.DebeziumLogicalConvertersTest

Tests run: 302, Failures: 0, Errors: 0, Skipped: 0
------------------------------------------------------------------------
BUILD SUCCESS
------------------------------------------------------------------------
Total time:  6.997 s
Finished at: 2025-08-12T13:49:14+02:00
------------------------------------------------------------------------
```

Claudenw commented:
I am confused about the changes to the logical converters.   It seems to me that we can use the setting for the Decimal or VariableScaleDecimal in the conversion for them and keep the conversion code contained.  I do see now that there is a chunk of functionality I did not think about.  Perhaps we can revisit that and try to merge these two options.

veliuenal commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/logicaltype/KafkaLogicalConverters.java:
```diff
@@ -80,14 +81,47 @@ public static class DecimalConverter extends LogicalTypeConverter {
       super(Decimal.LOGICAL_NAME,
           Schema.Type.BYTES,
-          LegacySQLTypeName.FLOAT);
+          LegacySQLTypeName.NUMERIC);
```
Updated for testing purposes.
We will keep the default type as FLOAT.

veliuenal commented:
Hey @Claudenw
Thank you for your comments.

Sorry, I didn’t get a chance to write the overview on Friday. Let me walk you through the changes file-by-file:

**BigQuerySinkConfig.java**
- Added a DecimalHandlingMode enum with config helpers (getDecimalHandlingMode, getVariableScaleDecimalHandlingMode).
- This allows case-insensitive selection of NONE, FLOAT, NUMERIC, or BIGNUMERIC modes for both fixed-scale and Debezium variable-scale decimals.

**BigQueryRecordConverter.java**

- Added new constructors and reworked convertLogical so the chosen handling mode is applied at runtime.
- Depending on the mode, decimals are either wrapped in {scale,value} structs, converted to double, or passed through as BigDecimal.
- Debezium converters are now only registered when actually needed.

**BigQuerySchemaConverter.java**
- Introduced convertDecimalField and convertVariableScaleDecimalField methods.
- These build BigQuery fields according to the selected handling mode and delegate precision/scale construction to logical converters.

**KafkaLogicalConverters.java**
- Updated DecimalConverter to inspect schema parameters, choose between NUMERIC and BIGNUMERIC based on precision/scale, and set those attributes on the resulting BigQuery field.
- Added an overridable getFieldBuilder hook in LogicalTypeConverter to support this.

I agree that the default type for DecimalConverter in KafkaLogicalConverters.java should be FLOAT.
I made a quick update to that part on Friday to test my changes, and now I’m extending them further.

In general, this is our need, based on my understanding of our discussions in the GitHub issue : https://github.com/Aiven-Open/bigquery-connector-for-apache-kafka/issues/76 ..
In light of the details shared here, could you let me know if these changes are on the right track for alignment with the Aiven platform?

Claudenw commented:
There is too much change to comment here.  If you want I will create a new PR with your changes and mine.

veliuenal commented:
>  If you want I will create a new PR with your changes and mine.
Sure, thanks @Claudenw

Claudenw commented:
This PR became the basis for pr #82 and code is included there.
